backend/routes/reservation.js

```diff
@@ -0,0 +1,101 @@
+import express from 'express';
+import bodyParser from 'body-parser';
+
+import connection from '../database';
+
+const router = express.Router();
+router.use(bodyParser.urlencoded({ extended: true }));
+
+// Retrieve a single reservation for a user to check their reservation.
+router.get('/single', (req, res) => {
+  const { reservationID } = req.query;
+
+  if (!reservationID) {
+    res.status(400).json({ error: 'reservation/single GET endpoint needs a reservationID query param' });
+    return;
+  }
+
+  connection.query(
+    `SELECT ` +
+      `ID, Date, Time, Notes, NumberOfGuests, TableID, RestaurantID, UserID FROM RESERVATION ` +
+      `WHERE ID = ? `,
+    [reservationID],
+    (error, results) => {
+      if (error) {
+        res.status(400).json({ error });
+        return;
+      }
+      res.json({ result: results });
+    }
+  );
+});
+
+// Retrieve all reservation from a single restaurant. Used by restaurant manager to see all reservations.
+router.get('/all', (req, res) => {
+  const { restaurantID } = req.query;
+
+  if (!restaurantID) {
+    res.status(400).json({ error: 'reservation/all GET endpoint needs a restaurantID query param' });
+    return;
+  }
+
+  connection.query(
+    `SELECT ID, Date, Time, Notes, NumberOfGuests, TableID, RestaurantID, UserID FROM RESERVATION WHERE RestaurantID = ?;`,
+    [restaurantID],
+    (error, results) => {
+      if (error) {
+        res.status(400).json({ error });
+        return;
+      }
+      res.json({ result: results });
+    }
+  );
+});
+
+// Add a new reservation when a user wants to book a table.
+router.post('/single', (req, res) => {
+  // Generate a random number capped at 2147483647 as this is the largest number the database can hold.
+  const reservationID = Math.floor(Math.random() * 2147483647);
+  const { date, time, notes, numberOfGuests, tableID, restaurantID, userID } = req.body;
+
+  if (!date || !time || !numberOfGuests || !tableID || !restaurantID || !userID) {
+    res.status(400).json({
+      error:
+        'reservation/single POST endpoint needs: date, time, numberOfGuests, tableID, restaurantID and userID body params',
+    });
+    return;
+  }
+
+  connection.query(
+    `INSERT INTO RESERVATION (ID, Date, Time, Notes, NumberOfGuests, TableID, RestaurantID, UserID) ` +
+      `VALUES (?, ?, ?, ?, ?, ?, ?, ?);`,
+    [reservationID, date, time, notes, numberOfGuests, tableID, restaurantID, userID],
+    error => {
+      if (error) {
+        res.status(400).json({ error });
+        return;
+      }
+      res.json({ result: 'Added single reservation', reservationID });
+    }
+  );
+});
+
+// Delete a reservation when a customer want to cancel a booking.
```

Comment:
since they can only cancel more than 1 hour before their booking, does this need to be checked? or this will be handled by the front end?

Reply:
I think the endpoint should handle the 1 hour time constraint and return an HTTP error status if this condition isn't met.

Reply:
@jacoballen4534 Are you able to update the function to consider the bookings time vs current time of canceling booking (this would give an HTTP error)? Either that or a new `cancelable status` field in booking table which the frontend can use.

Reply:
Can do this in a new PR

Reply:
Functionality for time restraint is implemented. Nice work Jacob